### PR TITLE
Getting Started doc and refactoring README

### PR DESCRIPTION
**Pull request by debasishg**
This PR -

* adds a Getting Started document 
* refactors `README` for `call-record-aggregator` and `taxi-ride`
* adds a Basic Concepts document

### CODE REVIEW

File: examples/call-record-aggregator/README.md [hunk 1]
```diff
@@ -1,20 +1,27 @@
-## Spark based Cloudflow Application
+## Akka and Spark based Cloudflow Application
 
 
 ### Problem Definition
 
+This application aggregates streaming data for phone call records. It's a mixed cloudflow application consisting of Akka and Spark based streamlets.
```
Comment (olofwalker) [inline examples/call-record-aggregator/README.md]:
```suggestion
This application aggregates streaming data for phone call records. It's a mixed Cloudflow application consisting of Akka and Spark-based streamlets.
```

File: examples/call-record-aggregator/README.md [hunk 2]
```diff
@@ -1,20 +1,27 @@
...
 
 
 ### Example Deployment example on GKE
 
-Steps:
+**Steps:**
 
-1) Make sure you have installed a gke cluster and you are running Cloudflow
-(check https://github.com/lightbend/cloudflow-installer for more).
+* Make sure you have installed a gke cluster with cloudflow running as per the [installation guide](https://github.com/lightbend/cloudflow-installer).
```
Comment (olofwalker) [inline examples/call-record-aggregator/README.md]:
```suggestion
* Make sure you have installed a GKE cluster with Cloudflow running as per the [installation guide](https://github.com/lightbend/cloudflow-installer).
```

File: examples/call-record-aggregator/README.md [hunk 3]
```diff
@@ -51,22 +60,22 @@ At the very end you should see the application image built and instructions for
 [info] You can deploy the application to a Kubernetes cluster using any of the the following commands:
 [info]  
 [info]   kubectl cloudflow deploy eu.gcr.io/my-awesome-project/call-record-aggregator:34-69082eb-dirty
-[info]   oc plugin cloudflow deploy eu.gcr.io/my-awesome-project/call-record-aggregator:34-69082eb-dirty
 [info]  
 [success] Total time: 24 s, completed Nov 8, 2019 3:44:21 PM
 ```
 
-4) Make sure you have the kubectl cloudflow plugin setup.
+* Make sure you have the `kubectl cloudflow` plugin setup.
```
Comment (olofwalker) [inline examples/call-record-aggregator/README.md]:
```suggestion
* Make sure you have the `kubectl cloudflow` plugin configured.
```

File: examples/taxi-ride/README.md [hunk 4]
```diff
@@ -53,21 +54,23 @@ taxi-ride-fare-processor-53e8cece-tm-bc9df878f-qnbsz    1/1     Running   0
...
+> **Note:** If you want to access the Flink job manager you can check which processor pod exposes the 8080 port and also port-forward to it.
+
+### Example Deployment on GKE
 
-Steps:
+**Steps:**
 
-1) Make sure you have installed a gke cluster and you are running Cloudflow
-(check https://github.com/lightbend/cloudflow-installer for more).
+* Make sure you have installed a gke cluster with cloudflow running as per the [installation guide](https://github.com/lightbend/cloudflow-installer).
```
Comment (olofwalker) [inline examples/taxi-ride/README.md]:
```suggestion
* Make sure you have installed a GKE cluster with Cloudflow running as per the [installation guide](https://github.com/lightbend/cloudflow-installer).
```

File: examples/taxi-ride/README.md [hunk 5]
```diff
@@ -102,19 +108,19 @@ At the very end you should see the application image built and instructions for
 [info] You can deploy the application to a Kubernetes cluster using any of the the following commands:
 [info]  
 [info]   kubectl cloudflow deploy eu.gcr.io/my-awesome-project/taxi-ride-fare:34-69082eb-dirty
-[info]   oc plugin cloudflow deploy eu.gcr.io/my-awesome-project/taxi-ride-fare:34-69082eb-dirty
 [info]  
 [success] Total time: 63 s, completed Nov 8, 2019 1:05:02 PM
 ```
 
-4) Make sure you have the kubectl cloudflow plugin setup.
+* Make sure you have the `kubectl cloudflow` plugin setup.
```
Comment (olofwalker) [inline examples/taxi-ride/README.md]:
```suggestion
* Make sure you have the `kubectl cloudflow` plugin configured.
```

File: docs/Basic Concepts.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
+# Cloudflow - Basic Concepts
```
Comment (skonto) [inline docs/Basic Concepts.md]: I think we should better avoid spaces in file names as it creates when you link to them via a url.

File: docs/Basic Concepts.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+This document gives you an overview of the main building blocks of a cloudflow application.
+
+## Application
+
+A Cloudflow application is a self-contained distributed system of data processing services connected together by data streams. A Cloudflow cluster can host multiple applications.
+
```
Comment (skonto) [inline docs/Basic Concepts.md]: Is `Cloudflow cluster` the term we want to use or its more about Cloudflow runtime?

Comment (skonto) [reply docs/Basic Concepts.md]: maybe s/`self-contained distributed system of data processing services connected together by data streams`/`a collection of data processing components connected via data streams"`

File: docs/Basic Concepts.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+![Application](images/apps-1.png?resize=400,400&classes=center)
+
+A Cloudflow application consists of the following basic building blocks:
+
+1. Streamlets
+2. Blueprint
+3. Deployed
```
Comment (skonto) [inline docs/Basic Concepts.md]: s/Deployed/`Deployed Application`

File: docs/Basic Concepts.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+We will go through each of the above in the following sections.
+
+![Application Building Blocks](images/apps-2.png?raw=true "Application Building Blocks")
+
+
+## Streamlets
+
+A streamlet is one of the building blocks of a Cloudflow application. Each streamlet represents a stage in the data transformation pipeline. Streamlets can have inlets and outlets that provide the conduit for the flow of data. Each inlet and outlet is typed and can handle data _of that specific type_ only.
```
Comment (skonto) [inline docs/Basic Concepts.md]: s/"Streamlets can have inlets and outlets that provide the conduit for the flow of data"/"Streamlets can have input endpoints called `inputs` and output endpoints called `outlets`...

Comment (skonto) [reply docs/Basic Concepts.md]: Pls underline somehow the introduced terms.

Comment (skonto) [inline docs/Basic Concepts.md]: s"...data transformation pipeline"/".... data transformation graph"?

File: docs/Getting Started.md [hunk 10]
```diff
@@ -0,0 +1,778 @@
+# Getting Started
```
Comment (skonto) [inline docs/Getting Started.md]: I think we should avoid spaces, in filenames see above.

File: docs/Getting Started.md [hunk 11]
```diff
@@ -0,0 +1,778 @@
...
+5. Deploy the application to the cluster using cloudflow CLI
+6. Show usage of CLI helpers like `status`, `list` etc.
+7. Show how to deal with http ingress and push some data to the application
+8. Verify that the application works
+9. Undeploy the application
+
+
+## Download and install the CLI
+
+cloudflow CLI can be downloaded from [lightbend public bintray repository](https://bintray.com/lightbend/cloudflow-cli). Download the version appropriate for your platform and install it in your local system. Please make sure you have the executable in your path with proper permission settings.
```
Comment (skonto) [inline docs/Getting Started.md]: s/cloudflow/Cloudflow

File: docs/Getting Started.md [hunk 12]
```diff
@@ -0,0 +1,778 @@
+# Getting Started
+
+This document introduces the basic steps needed to install all components of cloudflow and deploy and run a simple cloudflow application on the GKE cluster.
+
+We will discuss the following steps in sequence:
+
+1. Download and install cloudflow CLI
```
Comment (skonto) [inline docs/Getting Started.md]: I suggest we start with building the app and separate deployment steps eg. downloading the cli and running Cloudflow on GKE. This is a bit of a long list, let's split it to high level steps here.

File: docs/Getting Started.md (hunk 11 above)
Comment (skonto) [reply docs/Getting Started.md]: Should we provide some cli screenshots or use something like: https://github.com/faressoft/terminalizer

File: docs/Getting Started.md [hunk 13]
```diff
@@ -0,0 +1,778 @@
...
+9. Undeploy the application
+
+
+## Download and install the CLI
+
+cloudflow CLI can be downloaded from [lightbend public bintray repository](https://bintray.com/lightbend/cloudflow-cli). Download the version appropriate for your platform and install it in your local system. Please make sure you have the executable in your path with proper permission settings.
+
+## Develop a sample cloudflow application
+
+In this section we will develop a sample application that will feature the main components of cloudflow. It will be a simple hello-world style application, expressive enough to demonstrate the major features of cloudflow. It will not have many of the advanced features that cloudflow offers - the main idea is to give you a feel for what it takes to build a complete application and deploy it in running condition in a GKE cluster.
```
Comment (skonto) [inline docs/Getting Started.md]: s/cloudflow/Cloudflow

File: docs/Getting Started.md [hunk 14]
```diff
@@ -0,0 +1,778 @@
...
+    ),
+    name := "sensor-data",
+    organization := "com.lightbend",
+
+    scalaVersion := "2.12.10",
+    crossScalaVersions := Vector(scalaVersion.value)
+  )
+```
+
+cloudflow offers a few plugins that abstract quite a bit of boilerplates necessary to build a complete application. In this example we use the plugin `CloudflowAkkaStreamsApplicationPlugin` that provides you all building blocks of developing an Akka Streams based cloudflow application.
```
Comment (skonto) [inline docs/Getting Started.md]: s/"a few plugins"/"a several sbt plugins" we need to be specific here.

File: docs/Getting Started.md [hunk 15]
```diff
@@ -0,0 +1,778 @@
...
+
+> **Note:** You can use multiple plugins to develop an application that uses multiple runtimes (Akka, Spark, Flink etc.). For simplicity of this example we will be using only one.
+ 
+The above build script is standard Scala sbt - the only difference is the lugin which we provide as part of cloudflow.
```
Comment (skonto) [inline docs/Getting Started.md]: s/lugin/plugin

Comment (skonto) [reply docs/Getting Started.md]: Could you elaborate a bit more about the sbt plugins here and what they do. Maybe a diagram could help.

File: docs/Getting Started.md [hunk 16]
```diff
@@ -0,0 +1,778 @@
...
+
+### Schema first approach
+
+In cloudflow streamlets work with optional inputs and outputs that are statically typed. The types represent objects that the specific input / output can handle. The first step in application development is to encode the objects in the form of avro schema. cloudflow will generate appropriate classes corresponding to each schema.
```
Comment (skonto) [inline docs/Getting Started.md]: Let's point to some avro resource link for the user who is not familiar enough. Also we should point to the avro sbt plugin we use right?

File: docs/Basic Concepts.md (hunk 6 above)
Comment (skonto) [reply docs/Basic Concepts.md]: @debasishg I dont see how this document is linked to the rest. Is this pointed from somewhere?

File: docs/Getting Started.md [hunk 17]
```diff
@@ -0,0 +1,778 @@
...
+We will develop a simple pipeline that processes events from a wind turbine farm. The application will receive streaming data that will pass through the following transformations :
+
+* ingested by a streamlet (ingress)
+* processed by a streamlet to convert into a domain object (processor)
+* validated by a streamlet (splitter) that separates valid and invalid records
+* logged in to loggers to be checked at output (egress)
+
+Each of the above terminologies (ingress, processor, splitter and egress) are explained in the [Concepts Guide](https://blank.com). Here's an overview of the application pipeline architecture:
+
+![Application Pipeline](images/pipe.001.jpeg?raw=true "Application Pipeline")
```
Comment (skonto) [inline docs/Getting Started.md]: If you view how the doc is rendered there is a lot of space between the image and the net paragraph. Could you remove that?

File: docs/Getting Started.md [hunk 18]
```diff
@@ -0,0 +1,778 @@
...
+* **avro** : contains the avro schema of the domain objects
+* **blueprint** : contains the blueprint of the application in a file named `blueprint.conf`
+* **scala** : contains the source code of the application under the package name `sensordata`
+* **build.sbt** : the sbt build script
+
+### The sbt build script
+
+Here's a minimal version of the sbt build script:
+
+```
```
Comment (skonto) [inline docs/Getting Started.md]: I think you would want to add **build.sbt :** at the top as you did with the other files for consistency reasons.

File: docs/Basic Concepts.md (hunk 6 above)
Comment (debasishg) [reply docs/Basic Concepts.md]: @skonto You are correct - I will link this from the Getting Started doc. And I will change the names of the files to NOT have spaces in a separate PR.

File: docs/Basic Concepts.md (hunk 7 above)
Comment (debasishg) [reply docs/Basic Concepts.md]: done ..

File: docs/Basic Concepts.md (hunk 8 above)
Comment (debasishg) [reply docs/Basic Concepts.md]: Added a diagram for the deployed application

File: docs/Basic Concepts.md (hunk 9 above)
Comment (debasishg) [reply docs/Basic Concepts.md]: done ..

Comment (debasishg) [reply docs/Basic Concepts.md]: @skonto strictly speaking it's not a graph - it cannot have a cycle .. we can say `tree` but I think `pipelines` is more intuitive

File: docs/Getting Started.md (hunk 10 above)
Comment (debasishg) [reply docs/Getting Started.md]: will change all of them in a separate PR

File: docs/Getting Started.md (hunk 11 above)
Comment (debasishg) [reply docs/Getting Started.md]: This is just for the download part. I am assuming we will have a separate guide for the CLI ..

File: docs/Getting Started.md (hunk 13 above)
Comment (debasishg) [reply docs/Getting Started.md]: done ..

File: docs/Getting Started.md (hunk 14 above)
Comment (debasishg) [reply docs/Getting Started.md]: done ..

File: docs/Getting Started.md (hunk 15 above)
Comment (debasishg) [reply docs/Getting Started.md]: @skonto Intentionally I avoided the details. The purpose of this doc is that the user will follow it and have the application (described here) up and running. I guess we will have more details on the sbt plugins in developer docs. Once we have them we can refer to that from here.

File: docs/Getting Started.md (hunk 16 above)
Comment (debasishg) [reply docs/Getting Started.md]: done ..

File: docs/Getting Started.md (hunk 17 above)
Comment (debasishg) [reply docs/Getting Started.md]: @skonto I tried that a lot .. looks like it's beyond my capability .. what do I need to do to remove that gap ?

File: docs/Getting Started.md (hunk 18 above)
Comment (debasishg) [reply docs/Getting Started.md]: done ..

**Comment by maasg** [comment]
+1

**Pull request merged**